src/NBitcoin/ChainedBlock.cs

```diff
@@ -185,6 +184,24 @@ public override string ToString()
             return this.Height + "-" + this.HashBlock;
         }
 
+        /// <summary>
+        /// Finds the ancestor of this entry in the chain that matches the chained block header specified.
+        /// This can be used when you not sure whether the <paramref name="chainedBlockHeader"/> is on
+        /// the same chain as this block header or not.
+        /// <remarks>Will first search assuming on same chain and looking up height using skiplist, 
+        /// if doesn't match then will assume might be a different chain and then will linearly search 
+        /// for the hash.</remarks>
+        /// </summary>
+        /// <param name="chainedBlockHeader">The chained block header to search for.</param>
+        /// <returns>The chained block header or null if can't be found.</returns>
+        public ChainedBlock FindAncestor(ChainedBlock chainedBlockHeader)
+        {
+            ChainedBlock found = this.GetAncestor(chainedBlockHeader.Height);
+            if ((found != null) && (found.HashBlock == chainedBlockHeader.HashBlock))
+                return found;
+            return this.FindAncestorOrSelf(chainedBlockHeader.HashBlock);
```

Comment:
i don't think we need this, it will return null anyway, so we can just return 

if we have ChainedBlock, this means we have Height and we have Hash, so this means that if that chained block was ancestor of `this` it could only be at that Height, can't be anywhere else

and the method should probably still be named FindAncestorOrSelf because it should be possible to return self

Reply:
So I guess the difference with this method and GetAncestor() is this validates the hash. Will this work properly in the wallet sync case now? I had a hard time getting my head around what exactly was going on there.

Reply:
the wallet is receiving a new block from a source of blocks
the wallet maintains its own tip and the new block can be in several different states relative to this tip

it could be 
- the next block on the same chain
- the next block on a different chain (reorg)
- maybe even something completely different, like lower height block but on different chain due to reorg

but it does not really matter what is going on there, important thing is not to change the functionality, so we have two places we change code there:

1) `if (newTip.Height > this.walletTip.Height)`
here the new tip is higher than the wallet tip, so if we are on the same chain, we just need to sync the gap (if there is any)

we've had `ChainedBlock findTip = newTip.FAOS(this.walletTip.Height);` there, which attempted to go from newTip down and either walletTip was on the same chain or not, if not, it would go through the whole chain down to the genesis. but that's inefficient, we don't need to go lower than walletTip.Height to know that it is not there, so we need null here in such case

2) the else branch of 1) if with `ChainedBlock findTip = this.walletTip.GetAncestor(newTip.Height);`, so it is the same, just the higher tip is walletTip, so the very same applies here

in case 1) if the new tip is on the same chain as wallet, there is some code to sync the gap